### PR TITLE
Added method to compute over-the-wire CBOR encoded transaction size

### PR DESCRIPTION
**Pull request by crocodile-dentist**
# Description

A method for computing the true size of cbor encoded transactions is proposed which is useful and convenient for the upcoming changes to the diffusion layer.

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [ ] Self-reviewed the diff

### CODE REVIEW

File: eras/babbage/impl/src/Cardano/Ledger/Babbage/Tx.hs [hunk 1]
```diff
@@ -9,6 +9,8 @@ module Cardano.Ledger.Babbage.Tx (
 )
 where
 
+import Cardano.Ledger.Binary (EncCBOR (encCBOR))
+
```
Comment (lehins) [inline eras/babbage/impl/src/Cardano/Ledger/Babbage/Tx.hs]:
```suggestion
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Tx.hs [hunk 2]
```diff
@@ -12,6 +12,8 @@ module Cardano.Ledger.Conway.Tx (
 )
 where
 
+import Cardano.Ledger.Binary (EncCBOR (encCBOR))
+
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Tx.hs]:
```suggestion
```

File: libs/cardano-ledger-core/cardano-ledger-core.cabal [hunk 3]
```diff
@@ -1,6 +1,6 @@
 cabal-version:      3.0
 name:               cardano-ledger-core
-version:            1.14.1.0
+version:            1.14.2.0
```
Comment (lehins) [inline libs/cardano-ledger-core/cardano-ledger-core.cabal]: Addition of a new type class method is a breaking change according to PVP.
```suggestion
version:            1.15.0.0
```

File: libs/cardano-ledger-core/CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,9 @@
 # Version history for `cardano-ledger-core`
 
+## 1.14.2.0
+
+* Add `wireSizeTxF` to `EraTx` class
+
 ## 1.14.1.0
```
Comment (lehins) [inline libs/cardano-ledger-core/CHANGELOG.md]: New sections are never added in a normal PR. It is in one of the checkboxes that mast be marked in the PR. Also version bump is wrong
```suggestion
## 1.15.0.0

* Add `wireSizeTxF` to `EraTx` class
```

File: eras/shelley/impl/src/Cardano/Ledger/Shelley/Tx.hs [hunk 5]
```diff
@@ -164,7 +164,7 @@ auxDataShelleyTxL =
 {-# INLINEABLE auxDataShelleyTxL #-}
 
 -- | Size getter for `ShelleyTx`.
-sizeShelleyTxF :: Era era => SimpleGetter (ShelleyTx era) Integer
+sizeShelleyTxF :: (Era era, Num b) => SimpleGetter (ShelleyTx era) b
```
Comment (lehins) [inline eras/shelley/impl/src/Cardano/Ledger/Shelley/Tx.hs]: This is very dangerous, since it would be susceptible to overflows, depending on the type it is used with. I suggest adding a different helper lens for you use case
```suggestion
wireSizeShelleyTxF :: Era era => SimpleGetter (ShelleyTx era) Word32
wireSizeShelleyTxF = to $ \(TxConstr (Memo _ bytes)) -> 
  let n = SBS.length bytes
  in if n <= fromIntegral (maxBound :: Word32) 
       then fromIntegral @Int @Word32 n 
       else error $ "Impossible: Size of the transaction is too big: " ++ show n
{-# INLINEABLE size32ShelleyTxF #-}

sizeShelleyTxF :: Era era => SimpleGetter (ShelleyTx era) Integer
```

File: eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs [hunk 6]
```diff
@@ -236,13 +240,18 @@ auxDataAlonzoTxL = lens auxiliaryData (\tx txTxAuxData -> tx {auxiliaryData = tx
...
 -- | txsize computes the length of the serialised bytes
-sizeAlonzoTxF :: forall era. EraTx era => SimpleGetter (AlonzoTx era) Integer
-sizeAlonzoTxF =
+sizeAlonzoTxF ::
+  forall era b.
+  (EraTx era, Num b) =>
+  -- | encoding for CBOR size or mempool submission
+  (AlonzoTx era -> Encoding) ->
+  SimpleGetter (AlonzoTx era) b
+sizeAlonzoTxF enc =
```
Comment (lehins) [inline eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs]: Same as with `sizeShelleyTxF`., please revert the changes and add a specialized `wireSizeAlonzoTxF

**Review by lehins — CHANGES_REQUESTED** [review]
Integer overflows is one of the leading causes of vulnerabilities. We should do our absolute best in preventing or guarding against overflows, even if they seem impossible, hence are my suggestions. 

Other than that it looks good.

File: eras/alonzo/impl/CHANGELOG.md [hunk 7]
```diff
@@ -1,5 +1,9 @@
 # Version history for `cardano-ledger-alonzo`
 
+## 1.10.1.1
+
+* Added `wireSizeAlonzoTxF`
+
 ## 1.10.1.0
```
Comment (lehins) [inline eras/alonzo/impl/CHANGELOG.md]: `1.10.1.0` has not been released yet. Again, new sections in the changelogs are never added. Also addition of a new function is a minor version bump. Patch version bump is reserved for non-semantic or non visible changes (eg, tests, documentation etc.)

```suggestion
## 1.10.1.0

* Added `wireSizeAlonzoTxF`

```

File: eras/shelley/impl/CHANGELOG.md [hunk 8]
```diff
@@ -1,5 +1,9 @@
 # Version history for `cardano-ledger-shelley`
 
+## 1.12.3.1
+
+* Added `wireSizeShelleyTxF`
+
 ## 1.12.3.0
```
Comment (lehins) [inline eras/shelley/impl/CHANGELOG.md]: Ditto
```suggestion
## 1.12.3.0

* Added `wireSizeShelleyTxF`

```

File: libs/cardano-ledger-core/CHANGELOG.md [hunk 9]
```diff
@@ -1,5 +1,9 @@
 # Version history for `cardano-ledger-core`
 
+## 1.15.0.0
+
+* Add `wireSizeTxF` to `EraTx` class
+
 ## 1.14.1.0
```
Comment (lehins) [inline libs/cardano-ledger-core/CHANGELOG.md]: Ditto
```suggestion
## 1.15.0.0

* Add `wireSizeTxF` to `EraTx` class

```

File: libs/cardano-ledger-core/src/Cardano/Ledger/Core.hs [hunk 10]
```diff
@@ -153,8 +153,12 @@ class
 
   auxDataTxL :: Lens' (Tx era) (StrictMaybe (AuxiliaryData era))
 
+  -- | For estimations of impact on block space
```
Comment (lehins) [inline libs/cardano-ledger-core/src/Cardano/Ledger/Core.hs]:
```suggestion
  -- | For fee calculation and estimations of impact on block space
```

**Review by lehins — APPROVED** [review]
Looks great. Thank you. Couple of minor issues with changelogs. In case you are curious to learn our release and versioning process for the future you can find it here: https://github.com/IntersectMBO/cardano-ledger/blob/master/RELEASING.md

File: eras/alonzo/impl/CHANGELOG.md (hunk 7 above)
Comment (crocodile-dentist) [reply eras/alonzo/impl/CHANGELOG.md]: Oops, my bad, I've read releasing guide and looked at how some other commits look like before making my commits and I guess I didn't quite catch it.

File: libs/cardano-ledger-core/cardano-ledger-core.cabal (hunk 3 above)
Comment (crocodile-dentist) [reply libs/cardano-ledger-core/cardano-ledger-core.cabal]: I have a general question - I looked at chap and the latest released version of cardano-ledger-core I saw there is 1.13.2.0 and so that is why I stayed with 1.14 but bumped the minor version just to indicate a change. Presumably at the next release, the PVP rules would be satisfied since there was a major version bump that somebody else applied due to his/her braking changes, and which I leveraged for my patch. After releasing, it won't be immediately known why 1.13 was bumped to 1.14 (ie. mine or others' changes), but it won't matter because the major version was bumped in the aggregate.

File: libs/cardano-ledger-core/CHANGELOG.md (hunk 4 above)
Comment (crocodile-dentist) [reply libs/cardano-ledger-core/CHANGELOG.md]: Pardon my ignorance, but what do you mean when you say that new sections are never added - if a change is made which affects any major.minor.patch values then a new section must be added? I thought my changes warranted version bumps (please see my other comment) and so that is why I added these. For some internal change (eg. refactoring) I would reuse some section by prepending to it but I thought this was not the case here.

File: eras/allegra/impl/cardano-ledger-allegra.cabal [hunk 11]
```diff
@@ -67,7 +67,7 @@ library
         bytestring,
         cardano-crypto-class,
         cardano-ledger-binary ^>=1.3,
-        cardano-ledger-core >=1.13 && <1.15,
+        cardano-ledger-core >=1.13 && <1.16,
```
Comment (lehins) [inline eras/allegra/impl/cardano-ledger-allegra.cabal]: It is the lower bound that needs to be updated. `cardano-ledger-core-1.14.0.0` adds a function that `cardano-ledger-allegra` depends on. Which means it will no longer work with `cardano-ledger-core-1.13.x`
```suggestion
        cardano-ledger-core >=1.14 && <1.15,
```

File: eras/alonzo/impl/cardano-ledger-alonzo.cabal [hunk 12]
```diff
@@ -79,7 +79,7 @@ library
         cardano-ledger-allegra ^>=1.5,
         cardano-crypto-class,
         cardano-ledger-binary ^>=1.3,
-        cardano-ledger-core >=1.13 && <1.15,
+        cardano-ledger-core >=1.13 && <1.16,
```
Comment (lehins) [inline eras/alonzo/impl/cardano-ledger-alonzo.cabal]: Ditto
```suggestion
        cardano-ledger-core >=1.14 && <1.15,
```

File: eras/alonzo/test-suite/cardano-ledger-alonzo-test.cabal [hunk 13]
```diff
@@ -46,7 +46,7 @@ library
         bytestring,
         cardano-ledger-alonzo:{cardano-ledger-alonzo, testlib} >=1.7,
         cardano-ledger-binary:{cardano-ledger-binary, testlib} >=1.0,
-        cardano-ledger-core:{cardano-ledger-core, testlib} >=1.11 && <1.15,
+        cardano-ledger-core:{cardano-ledger-core, testlib} >=1.11 && <1.16,
```
Comment (lehins) [inline eras/alonzo/test-suite/cardano-ledger-alonzo-test.cabal]: This package does not need to change any dependencies
```suggestion
        cardano-ledger-core:{cardano-ledger-core, testlib} >=1.11 && <1.15,
```

File: eras/babbage/impl/cardano-ledger-babbage.cabal [hunk 14]
```diff
@@ -75,7 +75,7 @@ library
         cardano-ledger-allegra ^>=1.5,
         cardano-ledger-alonzo >=1.9 && <1.11,
         cardano-ledger-binary ^>=1.3,
-        cardano-ledger-core >=1.13 && <1.15,
+        cardano-ledger-core >=1.13 && <1.16,
```
Comment (lehins) [inline eras/babbage/impl/cardano-ledger-babbage.cabal]: Ditto
```suggestion
        cardano-ledger-core >=1.14 && <1.15,
```

File: eras/conway/impl/cardano-ledger-conway.cabal [hunk 15]
```diff
@@ -88,7 +88,7 @@ library
         cardano-ledger-allegra ^>=1.5,
         cardano-ledger-alonzo ^>=1.10,
         cardano-ledger-babbage ^>=1.8,
-        cardano-ledger-core >=1.13.2 && <1.15,
+        cardano-ledger-core >=1.13.2 && <1.16,
```
Comment (lehins) [inline eras/conway/impl/cardano-ledger-conway.cabal]:
```suggestion
        cardano-ledger-core >=1.14 && <1.15,
```

File: eras/mary/impl/cardano-ledger-mary.cabal [hunk 16]
```diff
@@ -76,7 +76,7 @@ library
         cardano-data ^>=1.2,
         cardano-ledger-allegra ^>=1.5,
         cardano-ledger-binary ^>=1.3,
-        cardano-ledger-core >=1.13 && <1.15,
+        cardano-ledger-core >=1.13 && <1.16,
```
Comment (lehins) [inline eras/mary/impl/cardano-ledger-mary.cabal]: Ditto
```suggestion
        cardano-ledger-core >=1.14 && <1.15,
```

File: eras/shelley/impl/cardano-ledger-shelley.cabal [hunk 17]
```diff
@@ -108,7 +108,7 @@ library
         cardano-data ^>=1.2.2,
         cardano-ledger-binary ^>=1.3,
         cardano-ledger-byron,
-        cardano-ledger-core >=1.13 && <1.15,
+        cardano-ledger-core >=1.13 && <1.16,
```
Comment (lehins) [inline eras/shelley/impl/cardano-ledger-shelley.cabal]: Ditto
```suggestion
        cardano-ledger-core >=1.14 && <1.15,
```

File: eras/shelley/test-suite/cardano-ledger-shelley-test.cabal [hunk 18]
```diff
@@ -78,7 +78,7 @@ library
         cardano-data >=1.2,
         cardano-ledger-binary:{cardano-ledger-binary, testlib} >=1.3,
         cardano-ledger-byron,
-        cardano-ledger-core:{cardano-ledger-core, testlib} >=1.11 && <1.15,
+        cardano-ledger-core:{cardano-ledger-core, testlib} >=1.11 && <1.16,
```
Comment (lehins) [inline eras/shelley/test-suite/cardano-ledger-shelley-test.cabal]: Ditto
```suggestion
        cardano-ledger-core:{cardano-ledger-core, testlib} >=1.11 && <1.15,
```

File: libs/cardano-ledger-api/cardano-ledger-api.cabal [hunk 19]
```diff
@@ -59,7 +59,7 @@ library
         cardano-ledger-babbage ^>=1.8.1,
         cardano-ledger-binary ^>=1.3,
         cardano-ledger-conway >=1.13 && <1.17,
-        cardano-ledger-core >=1.13.2 && <1.15,
+        cardano-ledger-core >=1.13.2 && <1.16,
```
Comment (lehins) [inline libs/cardano-ledger-api/cardano-ledger-api.cabal]: Ditto
```suggestion
        cardano-ledger-core >=1.13.2 && <1.15,
```

File: libs/cardano-ledger-core/cardano-ledger-core.cabal [hunk 20]
```diff
@@ -1,6 +1,6 @@
 cabal-version:      3.0
 name:               cardano-ledger-core
-version:            1.14.1.0
+version:            1.15.0.0
```
Comment (lehins) [inline libs/cardano-ledger-core/cardano-ledger-core.cabal]: My apologies. There was a mistake recently merged to master, that messed up the versioning process. @crocodile-dentist, it was thanks to you that I caught it. The latest released version is [`cardano-ledger-core-1.13.2.0`](https://github.com/IntersectMBO/cardano-haskell-packages/blob/main/_sources/cardano-ledger-core/1.13.2.0/meta.toml), which is also reflected with the tag: https://github.com/IntersectMBO/cardano-ledger/tree/cardano-ledger-core-1.13.2.0

In our versioning process it is important that developers:
* bump up the version when necessary in their PR, since they know much better if the change they introduce is breaking or not. For this reason at some point in the recent past we've already bumped `cardano-ledger-core-1.14.0.0` due to some breaking change.
* never add new section to the changelog, since that is done during a special post-release step. Recently we made a mistake of adding a new section in this commit: 067a1e574d1a6121dece06340b9670493a4f1939, which screwed everything up.

So, here is a PR that fixes that issue: #4532 

Which also means that this PR needs to bring down the version back to `1.14.0.0`. Again, sorry about the confusion.
```suggestion
version:            1.14.0.0
```

File: libs/cardano-ledger-core/CHANGELOG.md [hunk 21]
```diff
@@ -1,6 +1,8 @@
 # Version history for `cardano-ledger-core`
 
-## 1.14.1.0
+## 1.15.0.0
```
Comment (lehins) [inline libs/cardano-ledger-core/CHANGELOG.md]: See my comment on the version in the cabal file and changes in this PR: #4532 
```suggestion
## 1.14.0.0
```

**Review by lehins — APPROVED** [review]
We are currently having an issue with CI, somehow started running out of disk space, so don't worry about we are working on it.

File: libs/cardano-ledger-core/cardano-ledger-core.cabal (hunk 20 above)
Comment (crocodile-dentist) [reply libs/cardano-ledger-core/cardano-ledger-core.cabal]: It's all good, thanks for the help.
